### ISSUE TITLE
Multiple build-jobs can lead to exception due to throttling

### ISSUE THREAD

**Issue by fschaeffler** [zodern/mup-aws-beanstalk#16]
When multiple build jobs are running simultaneously, the API-limit for requests can get exceeded. IMO an exponential back-off (i.e., wait 2 seconds, then 4, 8, 16 up to a max such as 1024) would be an awesome feature.

*Log:*
```
00:37:12.982 => Setting up
00:37:13.705   Created Bucket
00:37:13.705 => Ensuring IAM Roles and Instance Profiles are setup
00:37:16.829   Created Beanstalk application
00:37:17.032 Building App Bundle Locally
00:37:57.765 => Archiving Bundle
00:38:02.152   10% Archived
00:38:03.131   20% Archived
00:38:04.085   30% Archived
00:38:04.856   40% Archived
00:38:05.768   50% Archived
00:38:06.612   60% Archived
00:38:07.532   70% Archived
00:38:08.399   80% Archived
00:38:09.299   90% Archived
00:38:09.960   100% Archived
00:38:11.726 => Uploading bundle
00:38:11.945   Uploaded 9%
00:38:11.947   Uploaded 18%
00:38:11.952   Uploaded 27%
00:38:12.844   Uploaded 36%
00:38:12.876   Uploaded 45%
00:38:13.279   Uploaded 54%
00:38:13.447   Uploaded 63%
00:38:13.684   Uploaded 72%
00:38:13.843   Uploaded 82%
00:38:14.701   Uploaded 91%
00:38:14.821   Uploaded 100%
00:38:14.821   Finishing upload. This could take a couple minutes
00:38:16.244 => Creating Version
00:38:16.501 => Configuring Beanstalk Environment
00:38:18.558  Created Environment
00:38:18.642 => Waiting for Beanstalk Environment to finish launching
00:42:45.412 => Configuring scaling
00:42:48.816 => Waiting for Beanstalk Environment to finish updating
00:42:55.308   Env Event: Updating environment *****'s configuration settings.
00:43:16.720   Env Event: Environment health has transitioned from Ok to Warning. 100.0 % of the requests are failing with HTTP 5xx. Initialization completed 42 seconds ago and took 3 minutes.
00:43:46.651 => Deploying new version
00:43:47.581 => Waiting for Beanstalk Environment to finish updating
00:44:15.430   Env Event: Environment health has transitioned from Warning to Severe. 100.0 % of the requests are failing with HTTP 5xx. ELB processes are not healthy on all instances. ELB health is failing or not available for all instances.
00:44:30.582   Env Event: Rolling with Additional Batch deployment policy enabled. Launching  a new batch of 1 additional instance(s).
00:46:16.050   Env Event: Added instance [i-******] to your environment.
00:46:23.265   Env Event: Batch 1: 1 EC2 instance(s) [i-******] launched. Deploying application version '1'.
00:46:25.495 { Throttling: Rate exceeded
00:46:25.495     at Request.extractError (/var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/protocol/query.js:47:29)
00:46:25.495     at Request.callListeners (/var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/sequential_executor.js:105:20)
00:46:25.495     at Request.emit (/var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
00:46:25.495     at Request.emit (/var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/request.js:683:14)
00:46:25.495     at Request.transition (/var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/request.js:22:10)
00:46:25.495     at AcceptorStateMachine.runTo (/var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/state_machine.js:14:12)
00:46:25.495     at /var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/state_machine.js:26:10
00:46:25.495     at Request.<anonymous> (/var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/request.js:38:9)
00:46:25.495     at Request.<anonymous> (/var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/request.js:685:12)
00:46:25.495     at Request.callListeners (/var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/sequential_executor.js:115:18)
00:46:25.495     at Request.emit (/var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
00:46:25.495     at Request.emit (/var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/request.js:683:14)
00:46:25.495     at Request.transition (/var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/request.js:22:10)
00:46:25.495     at AcceptorStateMachine.runTo (/var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/state_machine.js:14:12)
00:46:25.495     at /var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/state_machine.js:26:10
00:46:25.495     at Request.<anonymous> (/var/lib/jenkins/.nvm/versions/node/v8.9.1/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/request.js:38:9)
00:46:25.495   message: 'Rate exceeded',
00:46:25.495   code: 'Throttling',
00:46:25.495   time: 2018-01-24T15:58:44.648Z,
00:46:25.495   requestId: '******',
00:46:25.495   statusCode: 400,
00:46:25.495   retryable: true }
```

**Comment by zodern** [comment]
In b3fe265beacd90fd593b12e8f7469a398087b053 I increased the time between api requests when waiting for something to happen. 

I still want to add a back-off, and to better handle `Throttling` errors.

**Comment by zodern** [comment]
Thanks for finishing this @fschaeffler. Version 0.4.0 is now available with the changes.

**Comment by Discordius** [comment]
This still seems to happen to me on the latest version. Here is a log: 

```
=> Setting up
=> Ensuring IAM Roles and Instance Profiles are setup
Building App Bundle Locally
boilerplate-generator: updating npm dependencies -- combined-stream2...
=> Archiving Bundle
  10% Archived
  20% Archived
  30% Archived
  40% Archived
  50% Archived
  60% Archived
  70% Archived
  80% Archived
  90% Archived
  100% Archived
=> Uploading bundle
  Uploaded 3%
  Uploaded 7%
  Uploaded 10%
  Uploaded 14%
  Uploaded 17%
  Uploaded 21%
  Uploaded 24%
  Uploaded 28%
  Uploaded 31%
  Uploaded 35%
  Uploaded 38%
  Uploaded 42%
  Uploaded 45%
  Uploaded 49%
  Uploaded 52%
  Uploaded 56%
  Uploaded 59%
  Uploaded 63%
  Uploaded 66%
  Uploaded 70%
  Uploaded 73%
  Uploaded 77%
  Uploaded 80%
  Uploaded 84%
  Uploaded 87%
  Uploaded 91%
  Uploaded 94%
  Uploaded 96%
  Uploaded 100%
  Finishing upload. This could take a couple minutes
=> Creating Version
=> Configuring Beanstalk Environment
  Updated Environment
=> Waiting for Beanstalk Environment to finish updating
  Env Event: Immutable deployment policy enabled. Launching one instance with the new settings to verify health.
  Env Event: Created temporary auto scaling group awseb-e-2jukt5fcfp-immutable-stack-AWSEBAutoScalingGroup-4YKENRZL60XR.
  Env Event: Added instance [i-09c36911576b3a70b] to your environment.
  Env Event: Environment health has transitioned from Ok to Info. Configuration update in progress on 1 instance. 0 out of 1 instance completed (running for 2 minutes).
  Env Event: Waiting for instance(s) (i-09c36911576b3a70b) to pass health checks. 
  Env Event: Adding new instance(s) (i-09c36911576b3a70b) to the load balancer.
  Env Event: Test instance passed health checks. Launching remaining new instances.
in check exception
Setting new re-check interval to 4000ms
Throttling: Rate exceeded
    at Request.extractError (/home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/protocol/query.js:47:29)
    at Request.callListeners (/home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at Request.emit (/home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/request.js:683:14)
    at Request.transition (/home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/request.js:685:12)
    at Request.callListeners (/home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/sequential_executor.js:116:18)
    at Request.emit (/home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/request.js:683:14)
    at Request.transition (/home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/home/circleci/.meteor/packages/meteor-tool/.1.8.1-beta.0.16l1zj5.8ia9++os.linux.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/mup-aws-beanstalk/node_modules/aws-sdk/lib/request.js:38:9)
Exited with code 1
```

**Comment by Discordius** [comment]
It seems like it correctly sets the re-check interval, but then still crashes, so it isn't actually that useful.

**Comment by zodern** [comment]
I published version 0.6.2 to fix it.

**Comment by fschaeffler** [comment]
Sorry for the late response here. We created an internal fork of this repo, as we do all our deployments with the forked plugin. We needed more speed here because lots of changes were needed like e.g. optional IPv6-support, optional environment type for single-instance environments, optional custom DNS-servers, optional HTTP-headers for CSP, CDN for assets, auditd, chkroot and samhain on all instances, jumphost-definitions, ... and also the real fix for the API-throttling.

Not sure how the fix is now implemented, but in the fork it was solved with the following code in the file `src/aws.js`

```
  const MAX_BACKOFF_TIME_IN_MS = 3 * 60 * 1000;

  const options = {
    accessKeyId: auth.id,
    secretAccessKey: auth.secret,
    region: region || 'us-east-1',
    maxRetries: 25,
    retryDelayOptions: {
      customBackoff: (retryCount) => {
        const backoffTime = Math.min(((2 ** retryCount) * 1000), MAX_BACKOFF_TIME_IN_MS);
        if (retryCount >= 3) {
          console.log(`Retrying for the ${retryCount + 1}. time after a backoff-time of ${backoffTime / 1000}s`);
        }
        return backoffTime;
      }
    }
  };
```
